HPCC-20953 Add Connection close header for the last transaction

mayx opened the pull request:
- Add variable "shouldClose" to persistent connection handler methods
  getAvailable and notifySelected to indicate that the current transaction
  is the last one for the connection.
- HTTP server includes "Connection: close" in the response to the client
  for the last transaction.
- HTTP client includes "Connection: close" in the request for the last
  transaction

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck please review...

HPCCSmoketest commented:
Process of PR-12221, label: HPCC-20953-PersistClose starts now.
The reason of this test is: Did not build yet.
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 0ab867872bcc671b4c32e3545152aa5c5e9f6170
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |819 |819 |0 |
|test (thor) |744 |744 |0 |
|test (roxie) |893 |893 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   27 sec (00:00:27) |  192 sec (00:03:12) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1439 sec (00:23:59) |   19 sec (00:00:19) | 1697 sec (00:28:17) |

afishbeck commented on esp/platform/persistent.cpp:
```diff
@@ -206,6 +208,7 @@ class CPersistentHandler : implements IPersistentHandler, implements ISocketSele
         }
         else if (m_notify != nullptr)
         {
+            bool reachingQuota = false;
```
Should this be "reachedQuota"?

mayx replied:
Changed reachingQuota to reachedQuota (also changed another place "quotaReached" to "reachedQuota" for consistency) @afishbeck

afishbeck reviewed (commented):
@mayx one quite trivial comment.

mayx committed:
HPCC-20953 Add Connection close header for the last transaction

- Add variable "shouldClose" to persistent connection handler methods
  getAvailable and notifySelected to indicate that the current transaction
  is the last one for the connection.
- HTTP server includes "Connection: close" in the response to the client
  for the last transaction.
- HTTP client includes "Connection: close" in the request for the last
  transaction

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-12221, label: HPCC-20953-PersistClose starts now.
The reason of this test is: New commit.
Commit ID: 65bf4d89556eda48ccaa547f7ea1368454c41222
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 65bf4d89556eda48ccaa547f7ea1368454c41222
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |104 |104 |0 |0 |0 |34 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |819 |819 |0 |268 sec  (00:04:28) |
|test (thor) |744 |744 |0 |592 sec  (00:09:52) |
|test (roxie) |893 |893 |0 |350 sec  (00:05:50) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  192 sec (00:03:12) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1492 sec (00:24:52) |   20 sec (00:00:20) | 1747 sec (00:29:07) |

afishbeck reviewed (approved):
@mayx looks good.
